unit_tests/ex_suffix_expander.cc

```diff
@@ -9,13 +9,177 @@
 */
 
 #include <unordered_map>
+#include <cctype>    // isdigit, isspace, tolower
+#include <algorithm> // std::equal
+#include <iostream>  // debug (TODO: remove)
 
 #include "swoc/TextView.h"
 #include "swoc/swoc_file.h"
 #include "catch.hpp"
 
 using swoc::TextView;
 
+// TODO: some comments here stating the assumptions I've made
+
+// The Metric concept
+// A metric is a measurement system which can be computer storage, time duration
+// etc.
+/*
+struct Metric {
+  // the base (smallest) unit of the metric system
+  static constexpr TextView baseUnit = "";
+
+  // convert different units to a multiply of the base unit
+  static inline uintmax_t convertToBase(TextView const unit);
+
+  // canonicalize the spelling of the metric. e.g. both "b" and "byte" mean the
+  // same thing
+  // intentionally *not* taking a const unit b/c there might be use cases where
+  // the implementation wants to modify the view and then return it
+  static inline TextView canonicalize(TextView unit);
+};
+*/
+
+inline bool iequal(TextView const lhs, TextView const rhs) noexcept {
+  struct Pred {
+    bool operator()(char const l, char const r) noexcept { return tolower(l) == tolower(r); }
+  };
+  return std::equal(lhs.begin(), lhs.end(), rhs.begin(), rhs.end(), Pred{});
+}
+
+// namespace enclosing classes that meet the requirements of the Metric concept
+namespace metric
+{
+struct Storage {
+  static constexpr TextView B  = "B";
+  static constexpr TextView KB = "KB";
+  static constexpr TextView MB = "MB";
+  static constexpr TextView GB = "GB";
+  static constexpr TextView TB = "TB";
+  static constexpr TextView PB = "PB";
+
+  static constexpr TextView baseUnit = B;
+
+  static uintmax_t convertToBase(TextView const unit) {
+    if (unit == B) {
+      return static_cast<uintmax_t>(1);
+    } else if (unit == KB) {
+      return static_cast<uintmax_t>(1) << 10;
+    } else if (unit == MB) {
+      return static_cast<uintmax_t>(1) << 20;
+    } else if (unit == GB) {
+      return static_cast<uintmax_t>(1) << 30;
+    } else if (unit == TB) {
+      return static_cast<uintmax_t>(1) << 40;
+    } else if (unit == PB) {
+      return static_cast<uintmax_t>(1) << 50;
+    } else {
+      // TODO: error handling
+      return static_cast<uintmax_t>(0);
+    }
+  }
+
+  static inline TextView canonicalize(TextView unit) {
+    if (iequal(unit, "b") || iequal(unit, "byte")) {
+      return B;
+    } else if (iequal(unit, "k") || iequal(unit, "kb")) {
+      return KB;
+    } else if (iequal(unit, "m") || iequal(unit, "mb")) {
+      return MB;
+    } else if (iequal(unit, "g") || iequal(unit, "gb")) {
+      return GB;
+    } else if (iequal(unit, "t") || iequal(unit, "tb")) {
+      return TB;
+    } else if (iequal(unit, "p") || iequal(unit, "pb")) {
+      return PB;
+    } else {
+      // TODO: error handling
+      return B;
+    }
+  }
+};
+
+struct Duration {
+  static constexpr TextView SECOND = "second";
```

Comment:
Why not make this an `enum`?

Reply:
the use of these static constants are only in `convertToBase` and `canonicalize`. They are both directly consuming input from clients (parsed from a longer string), so the input params have to be strings.